app/src/policy_engine.h

```diff
@@ -0,0 +1,214 @@
+          {
+            obj.set(
+              "cty", ctx.new_string(std::get<std::string>(phdr.cty.value())));
+          }
+        }
+
+        if (phdr.x5chain.has_value())
+        {
+          auto x5_array = ctx.new_array();
+          size_t i = 0;
+
+          for (const auto& der_cert : phdr.x5chain.value())
+          {
+            auto pem = ccf::crypto::cert_der_to_pem(der_cert);
+            x5_array.set_at_index(i++, ctx.new_string(pem.str()));
+          }
+
+          obj.set("x5chain", std::move(x5_array));
+        }
+      }
+
+      // Extra Notary protected header parameters.
+      {
+        if (phdr.notary_signing_scheme.has_value())
+        {
+          obj.set(
+            "notary_signing_scheme",
+            ctx.new_string(phdr.notary_signing_scheme.value()));
+        }
+
+        if (phdr.notary_signing_time.has_value())
+        {
+          obj.set_int64(
+            "notary_signing_time", phdr.notary_signing_time.value());
+        }
+
+        if (phdr.notary_authentic_signing_time.has_value())
+        {
+          obj.set_int64(
+            "notary_authentic_signing_time",
+            phdr.notary_authentic_signing_time.value());
+        }
+
+        if (phdr.notary_expiry.has_value())
+        {
+          obj.set_int64("notary_expiry", phdr.notary_expiry.value());
+        }
+      }
+
+      return obj;
+    }
+
+    static inline bool apply_js_policy(
+      const PolicyScript& script,
+      const std::string& policy_name,
+      ClaimProfile claim_profile,
+      const scitt::cose::ProtectedHeader& phdr)
+    {
+      // Allow the policy to access common globals (including shims for
+      // builtins) like "console", "ccf.crypto"
+      ccf::js::CommonContext interpreter(ccf::js::TxAccess::APP_RO);
+
+      ccf::js::core::JSWrappedValue apply_func;
+      try
+      {
+        apply_func =
+          interpreter.get_exported_function(script, "apply", policy_name);
+      }
+      catch (const std::exception& e)
+      {
+        throw BadRequestError(
+          scitt::errors::PolicyError,
+          fmt::format("Invalid policy module: {}", e.what()));
+      }
+
+      auto profile_val = claim_profile_to_js_val(interpreter, claim_profile);
+      auto phdr_val = protected_headers_to_js_val(interpreter, phdr);
+
+      const auto result = interpreter.call_with_rt_options(
+        apply_func,
+        {profile_val, phdr_val},
+        std::nullopt,
+        ccf::js::core::RuntimeLimitsPolicy::NONE);
+
+      if (result.is_exception())
+      {
+        auto [reason, trace] = interpreter.error_message();
+
+        throw BadRequestError(
+          scitt::errors::PolicyError,
+          fmt::format(
+            "Error while applying policy: {}\n{}",
+            reason,
+            trace.value_or("<no trace>")));
+      }
+
+      // JS-style semantics - anything truthy becomes true
+      return result.is_true();
+    }
+  }
+
+  static inline bool run_policy_engine(
```

Comment:
Could it make sense to also pass the claim payload or the unprotected headers for cases where the policy should be applied there too?

Reply:
+1 for not passing the unprotected headers, they are by definition unsuitable for policy. The only thing I think could make sense is to pass a boolean flag indicating the presence of unprotected headers, so the policy can forbid that. My sense is that this should be forbidden by default though.

Payloads are authenticated, but could be large and CTS is meant to be agnostic to them. Inspection/certification if it takes place should be done by a separate confidential system, which attaches an attestation/policy receipt in protected headers, and CTS should verify those.

Payload inspection can be very expensive in terms of runtime, and large in terms of TCB footprint. It is best if it stays isolated for these reasons. CTS should focus on transparency.